geode-redis/src/main/java/org/apache/geode/redis/internal/RedisLockService.java

```diff
@@ -70,29 +71,38 @@ public AutoCloseableLock lock(ByteArrayWrapper key) throws InterruptedException
       throw new IllegalArgumentException("key cannot be null");
     }
 
+    KeyHashIdentifier lockKey = new KeyHashIdentifier(key.toBytes());
     Lock lock = new ReentrantLock();
-    Lock oldLock = map.putIfAbsent(key, lock);
+    Lock oldLock = weakReferencesTolocks.putIfAbsent(lockKey, lock);
+
     if (oldLock != null) {
       lock = oldLock;
-      // we need to get a reference to the actual key object so that the backing WeakHashMap does
-      // not clean it up.
-      for (ByteArrayWrapper keyInSet : map.keySet()) {
-        if (keyInSet.equals(key)) {
-          key = keyInSet;
-          break;
-        }
-      }
+
+      // we need to get a reference to the actual key object
+      // so that the backing WeakHashMap does not clean it up
+      // when garbage collection happens.
+      lockKey = getReferenceToLockKey(lockKey);
     }
 
     if (!lock.tryLock(timeoutMS, TimeUnit.MILLISECONDS)) {
-      throw new TimeoutException("Couldn't get lock for " + key.toString());
+      throw new TimeoutException("Couldn't get lock for " + lockKey.toString());
     }
 
-    return new AutoCloseableLock(key, lock);
+    return new AutoCloseableLock(lockKey, lock);
   }
 
-  int getMapSize() {
-    return map.size();
+  private KeyHashIdentifier getReferenceToLockKey(KeyHashIdentifier lockKey) {
+    KeyHashIdentifier referenceToLockKey = null;
+
+    synchronized (weakReferencesTolocks) {
+      for (KeyHashIdentifier keyInSet : weakReferencesTolocks.keySet()) {
+        if (keyInSet.equals(lockKey)) {
+          referenceToLockKey = keyInSet;
```

Comment:
you can just return `keyInSet` without having to set it onto `referenceToLockKey` and return `null` on line 105